editor/src/clj/editor/editor_extensions/runtime.clj

```diff
@@ -0,0 +1,370 @@
+       and code executes significantly faster.
+
+  Lua code evaluation is done using read / bind / invoke-{suspending,immediate}
+  cycle:
+    - read takes a string of Lua code and compiles it into a reusable prototype
+    - bind takes a prototype and creates a 0-arg Lua function (a closure), that,
+      when invoked, will evaluate the code in the runtime
+    - invoke-{suspending,immediate} will actually evaluate the code.
+  We split 'eval' into 'bind' and 'invoke-{suspending,immediate}' because there
+  are different evaluation semantics for invoke-suspending and invoke-immediate
+  that the users of the runtime need to choose."
+  (:refer-clojure :exclude [read])
+  (:require [cljfx.api :as fx]
+            [clojure.java.io :as io]
+            [dynamo.graph :as g]
+            [editor.editor-extensions.vm :as vm]
+            [editor.future :as future])
+  (:import [com.defold.editor.luart DefoldBaseLib DefoldCoroutine$Create DefoldCoroutine$Yield DefoldIoLib DefoldVarArgFn SearchPath]
+           [java.io File PrintStream Writer]
+           [java.nio.charset StandardCharsets]
+           [org.apache.commons.io.output WriterOutputStream]
+           [org.luaj.vm2 LoadState LuaError LuaFunction LuaValue]
+           [org.luaj.vm2.compiler LuaC]
+           [org.luaj.vm2.lib Bit32Lib CoroutineLib PackageLib PackageLib$lua_searcher PackageLib$preload_searcher StringLib TableLib]
+           [org.luaj.vm2.lib.jse JseMathLib JseOsLib]))
+
+(set! *warn-on-reflection* true)
+
+(deftype EditorExtensionsRuntime [lua-vm create resume status yield])
+
+(deftype Suspend [f args])
+
+(defprotocol SuspensionResult
+  (^:private deliver-result [suspend-result])
+  (^:private refresh-context? [suspend-result]))
+
+(extend-protocol SuspensionResult
+  nil
+  (deliver-result [x] (vm/->lua x))
+  (refresh-context? [_] false)
+  Object
+  (deliver-result [x] (vm/->lua x))
+  (refresh-context? [_] false)
+  Exception
+  (deliver-result [ex] (throw ex))
+  (refresh-context? [_] false))
+
+(defn and-refresh-context
+  "Given a value delivered from the suspendable function to the Lua runtime,
+  instruct the runtime to refresh the execution context before delivering the
+  result to the editor script"
+  [result]
+  (reify SuspensionResult
+    (deliver-result [_] (deliver-result result))
+    (refresh-context? [_] true)))
+
+(defn ->lua
+  "Convert Clojure data structure to Lua data structure"
+  ^LuaValue [x]
+  (vm/->lua x))
+
+(defn ->clj
+  "Convert Lua data structure to Clojure data structure
+
+  Might lock the runtime Lua VM while deserializing tables
+  Converts tables either to:
+  - vectors, if not empty and all keys are positive ints
+  - maps, otherwise. string keys are converted to keywords
+
+  Preserves LuaThread, File (from IoLib) and LuaFunction"
+  [^EditorExtensionsRuntime runtime lua-value]
+  (vm/->clj lua-value (.-lua-vm runtime)))
+
+(def ^:private ^:dynamic *execution-context* nil)
```

Comment:
Don't worry, it's not mutable, it's only re-bound when we refresh the evaluation context. It's also encapsulated in this namespace, and users of this namespace receive it as an argument to `lua-fn` or `suspendable-lua-fn`. This is approach is necessary because we don't want to pass evaluation context to the scripts — evaluation contexts have their own lifecycle, and if the script saves it and then re-uses it later, it will get stale results from `editor.get()` etc. I'll add a comment explaining it.